src/app/ReadClient.cpp

```diff
@@ -540,10 +540,27 @@ CHIP_ERROR ReadClient::ProcessEventReportIBs(TLV::TLVReader & aEventReportIBsRea
 
 CHIP_ERROR ReadClient::RefreshLivenessCheckTimer()
 {
+    CHIP_ERROR err = CHIP_NO_ERROR;
     CancelLivenessCheckTimer();
-    ChipLogProgress(DataManagement, "Refresh LivenessCheckTime with %d seconds", mMaxIntervalCeilingSeconds);
-    CHIP_ERROR err = InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->StartTimer(
-        System::Clock::Seconds16(mMaxIntervalCeilingSeconds), OnLivenessTimeoutCallback, this);
+    VerifyOrReturnError(mpExchangeCtx != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    System::Clock::Timeout timeout = System::Clock::Seconds16(mMaxIntervalCeilingSeconds);
+    if (mpExchangeCtx->IsUDPTransport())
+    {
+        auto margin = System::Clock::Milliseconds32(
+            (CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS + 1) *
+            (mpExchangeCtx->GetIdleRetransmitTimeoutTick() << CHIP_CONFIG_RMP_TIMER_DEFAULT_PERIOD_SHIFT));
```

Comment:
This logic should not be happening here at this layer - instead, we should be exposing a function on the exchange context that returns some/all of this information:

```
System::Clock::Milliseconds32 ExchangeContext::GetAckTimeout()
```

or something like that (@bzbarsky-apple might have better suggestions here).

Reply:
agree, just create ExchangeContext::GetAckTimeout, thanks